src/client_event.rs

```diff
@@ -0,0 +1,599 @@
+            });
+        });
+    }
+
+    fn receive(world: &mut World) {
+        world.resource_scope(|world, mut server: Mut<RepliconServer>| {
+            world.resource_scope(|world, registry: Mut<AppTypeRegistry>| {
+                world.resource_scope(|world, client_events: Mut<ClientEvents>| {
+                    let mut ctx = ServerReceiveCtx {
+                        registry: &registry.read(),
+                    };
+
+                    for client_event in &client_events.0 {
+                        let client_events = world
+                            .get_resource_mut_by_id(client_event.client_events_id)
+                            .expect("client events shouldn't be removed");
+
+                        // SAFETY: passed pointer was obtained from this event ID.
+                        unsafe {
+                            client_event.receive(&mut ctx, client_events.into_inner(), &mut server)
+                        };
+                    }
+                });
+            });
+        });
+    }
+
+    fn resend_locally(world: &mut World) {
+        world.resource_scope(|world, client_events: Mut<ClientEvents>| {
+            let world_cell = world.as_unsafe_world_cell();
+            for client_event in &client_events.0 {
+                // SAFETY: both resources mutably borrowed uniquely.
+                let (events, client_events) = unsafe {
+                    let events = world_cell
+                        .get_resource_mut_by_id(client_event.events_id)
+                        .expect("events shouldn't be removed");
+                    let client_events = world_cell
+                        .get_resource_mut_by_id(client_event.client_events_id)
+                        .expect("client events shouldn't be removed");
+                    (events, client_events)
+                };
+
+                // SAFETY: passed pointers were obtained from this event IDs.
+                unsafe {
+                    client_event.resend_locally(events.into_inner(), client_events.into_inner())
+                };
+            }
+        });
+    }
+
+    fn reset(world: &mut World) {
+        world.resource_scope(|world, client_events: Mut<ClientEvents>| {
+            for client_event in &client_events.0 {
+                let events = world
+                    .get_resource_mut_by_id(client_event.events_id)
+                    .expect("events shouldn't be removed");
+
+                // SAFETY: passed pointer was obtained from this event ID.
+                unsafe { client_event.reset(events.into_inner()) };
+            }
+        });
+    }
+}
+
+/// Registered client events.
+#[derive(Resource, Default)]
+struct ClientEvents(Vec<ClientEvent>);
+
+/// Type-erased functions and metadata for a registered client event.
+///
+/// Needed to process all events in a single system and call its functions without knowing the type.
+struct ClientEvent {
+    type_id: TypeId,
+    type_name: &'static str,
```

Comment:
These properties are only used in debug mode, so making them optional can save 24 bytes here.

Reply:
The compiler will remove them since it can detect if fields aren't used in release :)

Reply:
Oh didnt know that. Thanks.